driver-core/src/main/com/mongodb/client/model/expressions/MqlExpression.java

```diff
@@ -161,6 +170,72 @@ public BooleanExpression lte(final Expression lte) {
         return new MqlExpression<>(ast("$lte", lte));
     }
 
+    public BooleanExpression isBoolean() {
+        return new MqlExpression<>(ast("$type")).eq(of("bool"));
+    }
+
+    @Override
+    public BooleanExpression isBooleanOr(final BooleanExpression or) {
+        return this.isBoolean().cond(this, or);
+    }
+
+    public BooleanExpression isNumber() {
+        return new MqlExpression<>(astWrapped("$isNumber"));
+    }
+
+    @Override
+    public NumberExpression isNumberOr(final NumberExpression or) {
+        return this.isNumber().cond(this, or);
+    }
+
+    public BooleanExpression isString() {
+        return new MqlExpression<>(ast("$type")).eq(of("string"));
+    }
+
+    @Override
+    public StringExpression isStringOr(final StringExpression or) {
+        return this.isString().cond(this, or);
+    }
+
+    public BooleanExpression isDate() {
+        return ofStringArray("date", "timestamp").contains(new MqlExpression<>(ast("$type")));
+    }
+
+    @Override
+    public DateExpression isDateOr(final DateExpression or) {
+        return this.isDate().cond(this, or);
+    }
+
+    public BooleanExpression isArray() {
+        return new MqlExpression<>(astWrapped("$isArray"));
+    }
+
+    @SuppressWarnings("unchecked") // TODO
+    @Override
+    public ArrayExpression<Expression> isArrayOr(final ArrayExpression<? extends Expression> or) {
+        // TODO it seems that ArrEx<T> does not make sense here
+        return (ArrayExpression<Expression>) this.isArray().cond(this.assertImplementsAllExpressions(), or);
```

Comment:
I don't like how this method is turning out. Users evidently use `isArray`, possibly to handle fields where the value might be T vs an array of T (a pattern that our "match" has encouraged). But this method, which ensures that this is at least an array, does nothing (and can do nothing) to check the type of an array. This would be something like an assertion on the part of the user (much like `getIntegerField`), but mixing a validation with an assertion seems like it might be confusing.

I'd like to at least have another look at this at the end of Expressions, unless there is some idea to fix this up in this PR.

Reply:
If `ArrayExpression<T>` is needed, one can currently still achieve it as follows by honestly checking each element:
```java
ArrayExpression<BooleanExpression> booleans = ofBooleanArray(true, true)
    .isArrayOr(ofBooleanArray())
    .map(e -> e.isBooleanOr(of(false)));
```

We can even provide the `Expression.isIntegerOr` method, which will allow representing `ArrayExpression<NumberExpression>` as `ArrayExpression<IntegerExpression>` with the above approach. So I am proposing to introduce `Expression.isIntegerOr` (it can be implemented by checking whether `$type` is 32/64-bit integer).

Reply:
I agree with introducing the methods related to is-integer (in a followup PR).

Reply:
Leaving the two TODOs for the followup.

Reply:
When we have `Expression.isIntegerOr`, we should replace the cast in `array123.elementAt((IntegerExpression) of(0.0))` with `array123.elementAt(of(0.0).isIntegerOr(-1))`. Our tests should not test scenarios that do not use the API as intended, and casting is not intended.

Could you please link the PR where you will introduce the method? I will link the discussion there.

Reply:
https://github.com/mongodb/mongo-java-driver/pull/1052 Should be introduced here. I realize I also need to add tests for isIntegerOr in that PR.